### DISCUSSION TITLE
Backup options

### DISCUSSION THREAD

**Discussion by euu2021** [freeplane/freeplane#1290] [Ideas]
This recent data loss bug made me wonder if it wouldn't be relevant to have more backup options. It's possible that the data loss is only noticed after some weeks, so it would be useful to have backups for a longer time span.

I like how it works in Anki, which gives these options for users:
![image](https://github.com/freeplane/freeplane/assets/77707706/19c13a11-ce8f-4000-956f-8f6c587f2c83)

I just checked here, and my backups in FP go back to 5 days ago. In Anki, I have backups from almost 1 year ago.

**Comment by sifran-github** [comment]
Why not using an external backup software, that probably has many more features? and can backup many more things or just *.mm files.

**Reply by euu2021** [reply]
Are there good FOSS options out there?

The reality is that most users only get interested in the topic "backup" when it's too late 🤣🤣 After the disaster, they get VERY interested in knowing if the software creates backups.
So, maybe it's a good idea to provide it out of the box.

**Reply by macmarrum** [reply]
I didn't find any backup tool fitting my needs so I started to write a tool of my own:
https://github.com/macmarrum/rumar

It's a command-line tool (no GUI), with configuration in a toml file.

EDIT: please see https://github.com/macmarrum/rumar for instructions

**Reply by euu2021** [reply]
Interesting! I will try it

**Reply by macmarrum** [reply]
> Interesting! I will try it

I've added some more information in the home page https://github.com/macmarrum/rumar

**Reply by euu2021** [reply]
The compression is good 🎉🎉:
![image](https://github.com/freeplane/freeplane/assets/77707706/0f14cf4e-dec8-4a69-9136-74b19b5b2009)

Now, I'm wondering how to mimic the Anki system that I mentioned above. It's relevant because it keeps storage size under control. The idea is to automatically delete older backups, but leave, for example, one for each month.

**Reply by macmarrum** [reply]
> The compression is good 🎉🎉:

Cool. If you ever need more compression at the cost of time, this can be easily changed in settings.

> Now, I'm wondering how to mimic the Anki system that I mentioned above. It's relevant because it keeps storage size under control. The idea is to automatically delete older backups, but leave, for example, one for each month.

I will add such options soon.

**Comment by macmarrum** [comment]
@euu2021 Here's a preview version of [(dev) rumar.py](https://raw.githubusercontent.com/macmarrum/rumar/dev/src/rumar.py) with the functionality to remove old backup files.

I have tested it on my side. Could you please have a go at it?

Here's the latest README: https://github.com/macmarrum/rumar/tree/dev

The settings changed slightly, as did the command-line arguments (please see below).
```toml
# rumar.toml
backup_base_dir = 'c:\Users\Mac\Backup'

["My Documents"]
source_dir = 'c:\Users\Mac\Documents'
excluded_dirs_as_glob = ['/My Music', '/My Pictures', '/My Videos']
excluded_files_as_regex = ['/(desktop\.ini|thumbs\.db)$']
```

## How to use it

1. Install [Python](https://www.python.org/downloads/) (at least 3.9), if not yet installed
2. Download [rumar.py](https://raw.githubusercontent.com/macmarrum/rumar/main/src/rumar.py)
3. Create your `rumar.toml` settings in the same directory as `rumar.py` – see [settings example](#settings-example)
4. Open terminal in the directory containing `rumar.py`
5. If your installed Python version is below 3.11, run `python -m pip install tomli` to install the module [tomli](https://pypi.org/project/tomli/), if not yet installed
6. Run `python rumar.py list-profiles`; you should see your profile name printed in the console
7. Run `python rumar.py create --profile "My Documents"` to create a backup of the profile "My Documents"
8. Add this command to Task Scheduler or cron, to be run at an interval or each day/night

### How to sweep old backups

1. Run `python rumar.py sweep --profile "My Documents" --dry-run` and verify the files to be removed
2. Run `python rumar.py sweep --profile "My Documents"`
3. Add this command to Task Scheduler or cron, to be run at an interval or each day/night

Note: when --dry-run is used, file selection is run to count and select files to be removed but no files are actually deleted.

---

This version creates lots of debug log entries and a `rumar.sqlite` database. The last one is used by the new functionality to identify old backups to remove and can be useful in case of issues.

**Reply by macmarrum** [reply]
Please note some of the new settings:
* **number_of_daily_backups_to_keep** – 2 by default
* **number_of_weekly_backups_to_keep** – 14 by default
* **number_of_monthly_backups_to_keep** – 60 by default

**Reply by euu2021** [reply]
I'm trying to understand the meaning of, for example, `number_of_monthly_backups_to_keep: 60`. Which of those two options it mean:
- A) 60 backups will be kept each month
- B) 1 backup will be kept for each month of the last 60 months

In Anki, it's option B. In rumar, the high number that you chose by default made me wonder if I understood correctly.

**Reply by macmarrum** [reply]
This is a very good point. Yes, it's option A.

I've renamed the settings (in the latest [dev rumar.py](https://raw.githubusercontent.com/macmarrum/rumar/dev/src/rumar.py)) to make it clear:

- number_of_backups_per_day_to_keep
- number_of_backups_per_week_to_keep
- number_of_backups_per_month_to_keep

**Reply by euu2021** [reply]
Ok, I performed a test and noticed that I still don't understand the concept 🤣


- I created 3 backups in a matter of minutes.

- Then, I run this command: `python rumar.py sweep --profile "My Documents" `. I used those settings:
```
number_of_daily_backups_to_keep = 2
number_of_weekly_backups_to_keep = 14
number_of_monthly_backups_to_keep = 60
```
- Result: nothing changes. What I imagined that would happen: the oldest of the three backups would be deleted, because the daily limit of backups is 2.

So, my question is: what exactly is the meaning of `number_of_daily_backups_to_keep`.

To pose another situation: 
- let's say I've been using rumar for months, with those settings:
```
number_of_daily_backups_to_keep = 2
number_of_weekly_backups_to_keep = 14
number_of_monthly_backups_to_keep = 1
``` 
- how many backup files will I have if I look to a period of 1 month that already passed (for example, from May 1st to May 31)?

**Reply by macmarrum** [reply]
> I created 3 backups in a matter of minutes.
[...]
Result: nothing changes.

Ah, yes. This is a good thing, because it means the default setting to start sweeping only after 2 days is working as expected :smiley: 

→ **age_threshold_of_backups_to_sweep**: int = 2 &nbsp; &nbsp; _used by: sweep_\
→    only the backups which are older than the specified number of days are considered for removal

In other words, to run your test case, please change `age_threshold_of_backups_to_sweep` to 0.

Once this is done, you should see the oldest of the three backups deleted, because the daily limit of backups is 2.

**Reply by macmarrum** [reply]
Actually, you also need to change 
* number_of_weekly_backups_to_keep = 1
* number_of_monthly_backups_to_keep = 1

otherwise the weekly number will prevent deletion of your daily backups

**Reply by macmarrum** [reply]
> To pose another situation:
> 
> * let's say I've been using rumar for months, with those settings:
> 
> 
> ```
> number_of_daily_backups_to_keep = 2
> number_of_weekly_backups_to_keep = 14
> number_of_monthly_backups_to_keep = 1
> ```
> 
> * how many backup files will I have if I look to a period of 1 month that already passed (for example, from May 1st to May 31)?

It depends how many backups you created. Assuming you created 3 backups each day in May 2023 and none in June, after sweeping you'll end up with the following:

![2023-05-weeks](https://github.com/freeplane/freeplane/assets/7197131/fc81e162-1837-43d1-94c0-c1c1fe9807de)

* week 18: 14 backups out of 21 (7 days * 3 backups a day)
* week 19: 14 backups
* week 20: 14 backups
* week 21: 14 backups
* week 22: 9 backups out of 9 (3 days * 3 backups a day)

From README

* **number_of_backups_per_day_to_keep**: int = 2 &nbsp; &nbsp; _used by: sweep_\
  for each file, the specified number of backups per day is kept, if available, or more, to make weekly and/or monthly numbers\
  oldest backups are removed first
* **number_of_backups_per_week_to_keep**: int = 14 &nbsp; &nbsp; _used by: sweep_\
  for each file, the specified number of backups per week is kept, if available, or more, to make monthly numbers\
  oldest backups are removed first
* **number_of_backups_per_month_to_keep**: int = 60 &nbsp; &nbsp; _used by: sweep_\
  for each file, the specified number of backups per month is kept, if available\
  oldest backups are removed first

**Reply by euu2021** [reply]
I see that it's already packed with everything that I need. Thanks!

But, I'm still failing to make the sweep work 😅

I will report in the rumar repository.

**Comment by euu2021** [comment]
I just noticed that FP doesn't create backups for the media files. Maybe, it would be relevant to give a warning to the user. In Anki, for example, a warning is given inside the backup settings:
![image](https://github.com/freeplane/freeplane/assets/77707706/d23f4a0b-b9ba-4010-ab79-9e1f129b5bf5)
